README.md
[![Build Status](https://gitlab.com/canyan/canyan-tester/badges/master/pipeline.svg)](https://gitlab.com/canyan/canyan-tester/pipelines) [![codecov](https://codecov.io/gh/canyanio/canyan-tester/branch/master/graph/badge.svg)](https://codecov.io/gh/canyanio/canyan-tester) [![Docker pulls](https://img.shields.io/docker/pulls/canyan/canyan-tester.svg?maxAge=3600)](https://hub.docker.com/repository/docker/canyan/canyan-tester)

# Canyan Testing tool

Canyan Rating is an open source real-time highly scalable rating system. It is composed of an Agent Service, an API, and a Rating Engine.

The rating system is a critical component in any business, especially when real-time features are a strict requirement to ensure business continuity and congruence of transactions. Any compromise to availability, integrity, and authentication in the billing system makes a huge impact on the services provided.

Canyan aims to address these challenges with a cloud-native scalable solution, easily deployable and easily usable. It has been designed to work atomically ensuring the system status is always consistent, reproducible and coherent. Asynchronous processing of no real-time, consolidation events, prioritization, and time-boxed tasks provide the basics to ensure lightning-fast transaction processing without compromises.

Ease of use is addressed with comprehensive documentation, examples and high-quality software (see the test coverage badge).

Canyan Rating is designed as a microservice architecture and comprises [several repositories](https://github.com/canyanio). Its components are stateless and easily deployable via containers on-premises or in the cloud. 

This repository contains the Canyan coordinator and testing tool.

![Canyan logo](https://canyanio.github.io/rating-integration/canyan-logo.png) 


## Getting started

To start using Canyan Rating, we recommend that you begin with the Getting started
section in [the Canyan Rating documentation](https://canyanio.github.io/rating-integration/).


## Contributing

We welcome and ask for your contribution. If you would like to contribute to Canyan Rating, please read our guide on how to best get started [contributing code or documentation](https://canyanio.github.io/rating-integration/contributing/).


## License

Canyan is licensed under the GNU General Public License version 3. See
[LICENSE](https://canyanio.github.io/rating-integration/license/) for the full license text.


## Security disclosure

We take Canyan's security and our users trust very seriously.
If you believe you have found a security issue in Canyan, please responsibly
disclose by contacting us at [security@canyan.io](mailto:security@canyan.io).


## Running

Enter a Python 3 virtualenv and run the setup:

```
$ make setup
```

You can now run the test runner:

```
$ canyantester -t 1.2.3.4 -a http://api:8000 -e sipp sample.yaml
```

Where:

* **-t** is the target IP address
* **-a** is the address of the API server
* **-e** is the executable to run, defaults to `sipp`
* **sample.yaml** is the path to the canyantester configuration file


## Configuration file

`canyantester` accepts a YAML configuration file with the following root keys:

- setup
- teardown
- workers

### setup
Contain a list of actions to perform before the real testing. The parameters are the following:
- **type**: for now only `api` is implemented which makes a REST API call.

`api` supports the following configuration parameters:
* **uri**: the endpoint of the API where to make the request
* **method**: the http method to be used (defaults to `POST`)
* **store_response**: The name of the key for the `stored_responses` array used in other API calls for populating variables dependant of previous requests
* **payload**: the json payload to be sent (defaults to empty object)

#### Example using stored_responses
If you make an API call and use `store_response` value as `tenant` then in another API call you can use in the payload the value returned from the previous call like this:
```tenant_id: "{tenant.id}"```

### workers
Contains the list of workers to be created to perform the tests. 
There are two type of workers:
* sipp
* kamailio_xhttp

`sipp` supports the following configuration parameters:

* **number**: the number of instances to run (defaults to `1`)
* **repeat**: the number of times this worker should run (defaults to `1`)
* **timeout**: maximum number of seconds the worker can run before being killed, defaults to `None` (no timeout is enforced)
* **scenario**: path to the XML sipp scenario to run relative to the YAML configuration file
* **delay**: number of ms of delay before running the workers (defaults to `0`)
* **call_limit**: maximum number of concurrent active calls (`-l` parameter of sipp), defaults to `1`
* **call_number**: maximum number of calls (`-m` parameter of sipp) (defaults to `1`)
* **call_rate**: call rate increment (`-r` parameter of sipp) (defaults to `1`)
* **call\_rate\_period**: call rate period in ms (`-rp` parameter of sipp) (/(defaults to `1000`)
* **values**: key/value map of numerical or string values to be replaced in the XML tempalte.

The parameters which accept numerical values have support for random values expressed as follows:
```
delay:
  min: 1000
  max: 2000
```

Random values are calculated based on the random machine seed number printed by `canyantester` and configurabile through the `-s` option to make random tests reproducibles.


`kamailio_xhttp` supports the following configuration parameters:
* **uri**: the URI of the kamailio node and path where to send the request
* **delay**: number of seconds of delay before running the workers (defaults to `0`)
* **method**: the http method to be used (defaults to `POST`)
* **payload**: the json payload to be sent (defaults to empty object)


### check
This section contains a list of actions to perform for checking the correct data insertion during the workers process.
Useful for checking CDRs insertion and data consistency.
It implements only the `api` method.

The `api` type is the same as for `setup` except for the added `expected_response` which compares the API call response to the element specified in `expected_response`.

It also implements a `delay` in secods, useful for delays in data consolidation.


### teardown
Contain a list of actions to perform after the SIP tests.
It implements two methods:
* api
* kamailio_xhttp

The `api` type is the same as for `setup`. 
`kamailio_xhttp` is the same as for the `workers` but it's used without a delay and without threading.


## Connect with us

* Follow us on [Twitter](https://twitter.com/canyan_io). Please
  feel free to tweet us questions.
* Connect with us on [LinkedIN](https://www.linkedin.com/company/canyan/).
* Join us on [Slack](http://slack.canyan.io)
* Fork us on [Github](https://github.com/canyanio)
* Email us at [info@canyan.io](mailto:info@canyan.io)


CHANGELOG.md
# Changelog for canyan-tester

All notable changes to this module will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 1.0.0
- Initial release

## 1.1.0
- Add `check` section to run API calls after the workers are done to check and compare responses

## 1.2.0
- Better error and output handling
 

canyantester/tests/test_simple.py
import pytest  # type: ignore
import os

from canyantester import run_tester


def test_file_does_not_exist():
    with pytest.raises(FileNotFoundError):
        run_tester(config='/tmp/aaa.yaml')


def test_simple_command():
    with pytest.raises(FileNotFoundError):
        run_tester(
            executable='/bin/echo',
            config=os.path.join(
                os.path.dirname(__file__), 'scenarios', 'simple_command.yaml'
            ),
        )


canyantester/tests/test_canyantester.py
import pytest  # type: ignore
import unittest

from canyantester import run_tester


class canyantesterTest(unittest.TestCase):
    def test_canyantester_missing_parameters(self):
        with pytest.raises(FileNotFoundError):
            run_tester(config='/tmp/aaa.yaml')


canyantester/api.py
import json
import re
import requests
import socket

from dotty_dict import dotty  # type: ignore
from typing import Callable
from uuid import uuid4


RE_VARIABLE = re.compile(r'\{([^}\.]+)\.([^}]+)\}').search


def api_client(
    apiurl: str,
    config: dict,
    stored_responses: dict,
    verbose: bool = False,
    echo: Callable = print,
):
    uri = config.get('uri', '/')
    method = config.get('method', 'POST')
    payload = config.get('payload', None)
    expected_response = config.get('expected_response', None)
    stored_responses['random'] = {'uuid4': uuid4()}
    ipaddr = socket.gethostbyname(socket.gethostname())
    stored_responses['ipaddr'] = {'ip': ipaddr}

    variable_match = RE_VARIABLE(uri)
    if variable_match is not None:
        stored_response = dotty(stored_responses[variable_match.group(1)])
        uri = uri.replace(
            variable_match.group(0), str(stored_response[variable_match.group(2)])
        )

    if payload:
        for k, v in payload.items():
            if not isinstance(v, str):
                continue
            variable_match = RE_VARIABLE(v)
            if variable_match is None:
                continue
            stored_response = dotty(stored_responses[variable_match.group(1)])
            payload[k] = v.replace(
                variable_match.group(0), str(stored_response[variable_match.group(2)])
            )
    if method == 'POST':
        response = requests.post("%s%s" % (apiurl, uri), json=payload)
        if response.status_code != 200:
            echo("HTTP response code: %d" % response.status_code)
            echo("Payload: %s" % payload)
            echo("Response: %s" % response.text)
            raise RuntimeError()
        data = json.loads(response.text)
        if expected_response:
            echo("Comparing results...")
            if data != json.loads(expected_response):
                echo("NO Match! exiting...")
                echo("Expected: %s" % expected_response)
                echo("Response: %s" % data)
                raise RuntimeError()
            else:
                echo("OK: Results match!")
                if verbose:
                    echo("Expected: %s" % expected_response)
                    echo("Response: %s" % data)
                return data
        elif isinstance(data, dict):
            return data
        else:
            return None
    elif method == 'DELETE':
        response = requests.delete("%s%s" % (apiurl, uri))
        if response.status_code != 200:
            echo(response)
            raise RuntimeError()
        if verbose:
            data = json.loads(response.text)
            echo('Response: ')
            echo(data)
        return response.status_code == 200


canyantester/sipp.py
import os
import re
import subprocess
import time

from dotty_dict import dotty  # type: ignore

from .defaults import TEMPLATE_XML as DEFAULT_TEMPLATE_XML
from .utils import get_int_from_config
from .worker import Worker


class SippWorker(Worker):

    TEMPLATE_XML = DEFAULT_TEMPLATE_XML
    RE_VARIABLE = re.compile(r'\{([^}\.]+)\.([^}]+)\}').search

    def __init__(
        self,
        worker_id,
        config,
        target=None,
        executable=None,
        directory=None,
        basedir=None,
        log=print,
        stored_responses=None,
        verbose=False,
    ):
        super(SippWorker, self).__init__()
        self._worker_id = worker_id
        self._config = config
        self._directory = directory
        self._basedir = basedir
        self._log = log
        self._exception = None
        self._target = target
        self._executable = executable
        self._stored_responses = stored_responses
        self._verbose = verbose
        self._values = self._config.get('values', {})
        self._filename_xml = os.path.join(
            self._directory, 'sipp-%s.xml' % self._worker_id
        )
        self._delay = get_int_from_config(self._config, 'delay', 0)
        self._output = None
        self._args = []

    def setup(self):
        for k, v in self._values.items():
            if not isinstance(v, str):
                continue
            variable_match = self.RE_VARIABLE(v)
            if variable_match is None:
                continue
            stored_response = dotty(self._stored_responses[variable_match.group(1)])
            self._values[k] = v.replace(
                variable_match.group(0), str(stored_response[variable_match.group(2)])
            )
        values = {'basedir': self._basedir, 'target': self._target}
        for key in self._values.keys():
            if isinstance(self._values[key], str):
                values[key] = self._values[key]
            elif isinstance(self._values[key], (list, tuple)):
                worker_positional_id = int(self._worker_id.split('_', 1)[1])
                values[key] = self._values[key][
                    worker_positional_id % len(self._values[key])
                ]
            else:
                values[key] = get_int_from_config(self._values, key, None)
        if self._config.get('scenario'):
            template_xml = (
                open(os.path.join(self._basedir, self._config.get('scenario'))).read()
                % values
            )
        else:
            template_xml = self.TEMPLATE_XML % values
        with open(self._filename_xml, 'wb') as f:
            f.write(template_xml.encode('utf-8'))
        self._args = [
            self._executable,
            self._target,
            '-sf',
            self._filename_xml,
            '-l',
            str(get_int_from_config(self._config, 'call_limit', 1)),
            '-r',
            str(get_int_from_config(self._config, 'call_rate', 1)),
            '-rp',
            str(get_int_from_config(self._config, 'call_rate_period', 1000)),
            '-m',
            str(get_int_from_config(self._config, 'call_number', 1)),
        ]
        extra_args = self._config.get('extra_args')
        if isinstance(extra_args, (tuple, list)):
            self._args.extend(extra_args)
        elif isinstance(extra_args, str):
            self._args.extend(extra_args.split())
        self._output = None

    def runner(self):
        if os.path.exists(self._filename_xml.replace('.xml', '.log')):
            os.unlink(self._filename_xml.replace('.xml', '.log'))
        exit_codes = []
        with open(self._filename_xml.replace('.xml', '.log'), 'ab') as log:
            if self._delay:
                time.sleep(self._delay / 1000.0)
            for _ in range(get_int_from_config(self._config, 'repeat', 1)):
                self._log(
                    "[%s] %s (delay = %s)"
                    % (self._worker_id, " ".join(self._args), self._delay)
                )
                try:
                    p = subprocess.run(
                        args=self._args,
                        check=True,
                        stdout=log,
                        stderr=log,
                        timeout=self._config.get('timeout', None),
                    )
                except subprocess.CalledProcessError as e:
                    exit_codes.append(e.returncode)
                except subprocess.TimeoutExpired:
                    exit_codes.append(-1)
                else:
                    exit_codes.append(p.returncode)
        self._exit_codes = exit_codes

    def debug(self):
        non_zero_exit_codes = list(filter(lambda x: x != 0, self._exit_codes))
        self._exit_status = (
            non_zero_exit_codes[0]
            if len(non_zero_exit_codes) > self._config.get('max_errors', 0)
            else 0
        )
        self._log(
            "[%s] runs = %s, errors = %s, exit codes = %s, input = %s, output = %s"
            % (
                self._worker_id,
                len(self._exit_codes),
                len(non_zero_exit_codes),
                ', '.join(map(str, self._exit_codes)),
                self._filename_xml,
                self._filename_xml.replace('.xml', '.log'),
            )
        )


canyantester/defaults.py
TEMPLATE_XML = """<?xml version="1.0" encoding="iso-8859-2" ?>
<!DOCTYPE scenario SYSTEM "sipp.dtd">

<scenario name="Call with enabled profile and good credential 5calls x seconds max peak 300 simultaneous calls">

    <send retrans="500">
          <![CDATA[

                INVITE sip:[field3]@[remote_ip]:[remote_port] SIP/2.0
                Via: SIP/2.0/[transport] [local_ip]:[local_port];branch=[branch]
                From: sipp <sip:[field0]@[field1]>;tag=[call_number]
                To: <sip:[field3]@[field1]:[remote_port]>
                Call-ID: [call_id]
                CSeq: [cseq] INVITE
                Contact: sip:[field0]@[local_ip]:[local_port]
                Max-Forwards: 70
                Content-Type: application/sdp
                Content-Length: [len]

                v=0
                o=user1 53655765 2353687637 IN IP[local_ip_type] [local_ip]
                s=-
                c=IN IP[media_ip_type] [media_ip]
                t=0 0
                m=audio [media_port] RTP/AVP 8
                a=rtpmap:8 PCMA/8000

          ]]>
    </send>

    <recv response="100" optional="true"></recv>
    <recv response="407" auth="true" rrs="true"></recv>

    <send>
    <![CDATA[

          ACK sip:[field3]@[remote_ip]:[remote_port] SIP/2.0
          [last_Via:]
          [last_From:]
          [last_To:]
          [last_Call-ID:]
          CSeq: [cseq] ACK
          Contact: <sip:[local_ip]:[local_port];transport=[transport]>
          Max-Forwards: 60
          Content-Length: 0

    ]]>
    </send>


    <send retrans="500">
          <![CDATA[

                INVITE sip:[field3]@[remote_ip]:[remote_port] SIP/2.0
                Via: SIP/2.0/[transport] [local_ip]:[local_port];branch=[branch]
                From: sipp <sip:[field0]@[field1]>;tag=[pid]sippTag01[call_number]
                To: <sip:[field3]@[field1]:[remote_port]>
                Call-ID: [call_id]
                CSeq: [cseq] INVITE
                Contact: <sip:[local_ip]:[local_port];transport=[transport]>
                [field2]
                Max-Forwards: 50
                Content-Type: application/sdp
                Content-Length: [len]

                v=0
                o=user1 53655765 2353687637 IN IP[local_ip_type] [local_ip]
                s=-
                c=IN IP[media_ip_type] [media_ip]
                t=0 0
                m=audio [media_port] RTP/AVP 8
                a=rtpmap:8 PCMA/8000

          ]]>
    </send>

    <recv response="100" optional="true"></recv>
    <recv response="183" optional="true"></recv>
    <recv response="180" optional="true"></recv>
    <recv response="200" rrs="true"></recv>

    <send>
          <![CDATA[

          ACK [next_url] SIP/2.0
          [last_Via:]
          [routes]
          [last_From:]
          [last_To:]
          Call-ID: [call_id]
          CSeq: [cseq] ACK
          Contact: sip:[field0]@[local_ip]:[local_port]
          Max-Forwards: 40
          Content-Length: 0

          ]]>
    </send>

    <pause milliseconds="%(call_duration)d" />

    <send retrans="500">
          <![CDATA[

          BYE [next_url] SIP/2.0
          [last_Via:]
          [routes]
          [last_From:]
          [last_To:]
          [last_Call-ID:]
          CSeq: [cseq] BYE
          Contact: <sip:[local_ip]:[local_port];transport=[transport]>
          Max-Forwards: 30
          Content-Length: 0

          ]]>
    </send>

    <!-- The 'crlf' option inserts a blank line in the statistics report. -->
    <recv response="200" crlf="true">
    </recv>

    <!-- definition of the response time repartition table (unit is ms)       -->
    <ResponseTimeRepartition value="10, 20, 30, 40, 50, 100, 150, 200"/>

    <!-- definition of the call length repartition table (unit is ms)             -->
    <CallLengthRepartition value="10, 50, 100, 500, 1000, 5000, 10000"/>

</scenario>
"""


canyantester/__init__.py
import click
import os
import random
import requests
import shutil
import signal
import tempfile
import threading
import time

from yaml import load


try:
    from yaml import CLoader as Loader
except ImportError:
    from yaml import Loader  # type: ignore

from .utils import generate_random_seed, get_int_from_config
from .api import api_client
from .sipp import SippWorker


def print_version(ctx, _, value):
    if not value or ctx.resilient_parsing:
        return
    click.echo('Version 1.2.0')
    ctx.exit()


@click.command()
@click.option(
    '--version', is_flag=True, callback=print_version, expose_value=False, is_eager=True
)
@click.argument('config', type=click.File('rb'))
@click.option(
    '-t',
    '--target',
    default='sbc:5060',
    type=click.STRING,
    required=False,
    help='IP address of the SIP server to use as target',
)
@click.option(
    '-e',
    '--executable',
    type=click.STRING,
    default="sipp",
    help='Command to exec for running sipp',
)
@click.option(
    '-d',
    '--directory',
    type=click.Path(exists=True),
    default=None,
    help='Working directory, if not specified a temporary directory is created.',
)
@click.option(
    '-s',
    '--seed',
    type=click.INT,
    default=None,
    help='Initialize the Python random machine with this seed value.',
)
@click.option('-a', '--apiurl', default='http://api:8000', show_default=True)
@click.option(
    "--no-setup",
    is_flag=True,
    default=False,
    help="Skip setup step in yaml file",
    hidden=False,
)
@click.option(
    "--no-teardown",
    is_flag=True,
    default=False,
    help="Skip teardown step in yaml file",
    hidden=False,
)
@click.option("--verbose", is_flag=True, default=False, hidden=False)
def canyantester(
    config,
    target,
    executable,
    directory,
    seed,
    apiurl=None,
    cache_accounts=None,
    no_setup=False,
    no_teardown=False,
    verbose=False,
):
    """
    Test coordinator and runner, it allows to run real-world scenarios and stress tests
    coordinating multiple sipp instances.
    """
    try:
        run_tester(
            config=config,
            target=target,
            executable=executable,
            directory=directory,
            seed=seed,
            apiurl=apiurl,
            cache_accounts=cache_accounts,
            no_setup=no_setup,
            no_teardown=no_teardown,
            verbose=verbose,
            echo=click.echo,
        )
    except RuntimeError as e:
        raise click.Abort(str(e))


def run_tester(
    config,
    target='sbc:5060',
    executable='sipp',
    directory=None,
    seed=None,
    apiurl=None,
    cache_accounts=None,
    no_setup=False,
    no_teardown=False,
    verbose=False,
    echo=print,
):
    if seed is None:
        seed = generate_random_seed()

    remove_directory_when_done = False
    if directory is None:
        directory = tempfile.mkdtemp()
        remove_directory_when_done = True

    echo("Setting the random seed: %s" % seed)
    random.seed(seed)

    echo("Using temporary directory: %s" % directory)
    echo("Using executable: %s" % executable)
    echo("Target: %s" % target)

    if isinstance(config, str):
        config = open(config, 'rb')

    config_data = load(config, Loader=Loader)
    basedir = config_data.setdefault('basedir', os.path.dirname(config.name))
    stored_responses = {}

    setup = config_data.get('setup', None)
    if not no_setup and setup is not None:
        for i, setup_config in enumerate(setup):
            if setup_config.get('type', 'api') == 'api':
                store_response = setup_config.get('store_response', None)
                response = APIcall(apiurl, setup_config, stored_responses, verbose)
                if store_response:
                    stored_responses[store_response] = response
    else:
        echo("Skipping setup...")

    def _do_check():
        do_check(config_data, apiurl, stored_responses, verbose, echo=echo)

    def _do_teardown():
        do_teardown(
            config_data, no_teardown, apiurl, stored_responses, verbose, echo=echo
        )

    signal.signal(signal.SIGINT, _do_teardown)

    try:
        workers = config_data.get('workers')
        if workers is None:
            echo("No worker has been defined, quit!")
            raise RuntimeError()

        testers = []
        other_testers = []
        for i, worker_config in enumerate(workers):
            number_of_workers = get_int_from_config(worker_config, 'number', 1)

            for j in range(number_of_workers):
                if worker_config.get('type', 'sipp') == 'sipp':
                    tester = SippWorker(
                        worker_id="%06d_%06d" % (i, j),
                        config=worker_config,
                        target=target,
                        executable=executable,
                        directory=directory,
                        basedir=basedir,
                        log=echo,
                        stored_responses=stored_responses,
                        verbose=verbose,
                    )
                    tester.setup()
                    testers.append(tester)
                elif worker_config.get('type', None) == 'kamailio_xhttp':
                    thread = threading.Thread(
                        target=kamailioXHTTP, args=(worker_config, verbose)
                    )
                    other_testers.append(thread)

        echo("\nStarting workers:")
        threads = []
        for tester in testers:
            t = threading.Thread(target=tester)
            threads.append(t)
            t.start()

        for t in threads:
            t.join()

        for t in other_testers:
            t.start()

        for t in other_testers:
            t.join()

        echo("\nWorkers' results:")
        error = False
        for tester in testers:
            tester.debug()
            exit_status = tester.get_exit_status()
            if exit_status != 0:
                error = True
            tester.teardown()

        if error:
            raise RuntimeError("\nError detected, aborted!")
        else:
            if remove_directory_when_done:
                shutil.rmtree(directory)
            echo("\nDone!")
    finally:
        _do_check()
        _do_teardown()


def do_delay(config):
    delay = config.get('delay', 0)
    if delay:
        time.sleep(delay)


def APIcall(apiurl, config, stored_responses, verbose, echo=print):
    return api_client(
        apiurl=apiurl,
        config=config,
        stored_responses=stored_responses,
        verbose=verbose,
        echo=echo,
    )


def kamailioXHTTP(config, verbose, echo=print):
    do_delay(config)
    if config.get('method', "POST") == 'POST':
        uri = config.get('uri', "")
        payload = config.get('payload', {})
        response = requests.post(uri, json=payload)
        echo("HTTP response code: %d" % response.status_code)
        if response.status_code != 200:
            echo("Payload: %s" % payload)
            echo("Response: %s" % response.text)
            raise RuntimeError()
        if verbose:
            echo("Payload: %s" % payload)
            echo("Response: %s" % response.text)


def do_check(config_data, apiurl, stored_responses, verbose, echo=print):
    check = config_data.get('check', None)
    if check is not None:
        echo("Starting check process...")
        for _, check_config in enumerate(check):
            do_delay(check_config)
            if check_config.get('type', 'api') == 'api':
                store_response = check_config.get('store_response', None)
                response = APIcall(
                    apiurl, check_config, stored_responses, verbose, echo=echo
                )
                if store_response:
                    stored_responses[store_response] = response


def do_teardown(
    config_data, no_teardown, apiurl, stored_responses, verbose, echo=print
):
    echo("Starting teardown process...")
    teardown = config_data.get('teardown', None)
    if not no_teardown and teardown is not None:
        for _, teardown_config in enumerate(teardown):
            if teardown_config.get('type', 'api') == 'api':
                store_response = teardown_config.get('store_response', None)
                response = APIcall(apiurl, teardown_config, stored_responses, verbose)
                if store_response:
                    stored_responses[store_response] = response
            elif teardown_config.get('type', None) == 'kamailio_xhttp':
                kamailioXHTTP(teardown_config, verbose)
                time.sleep(5)

    else:
        echo("Skipping teardown procedure...")
